lib/ddtrace/configuration/components.rb

```diff
@@ -54,28 +54,116 @@ def build_tracer(settings, agent_settings)
+              sampler
+            else
+              ensure_priority_sampling(sampler)
+            end
+          elsif settings.tracer.priority_sampling == false
+            Sampling::RuleSampler.new(
+              rate_limit: settings.sampling.rate_limit,
+              default_sample_rate: settings.sampling.default_rate
+            )
+          else
+            PrioritySampler.new(
               base_sampler: AllSampler.new,
               post_sampler: Sampling::RuleSampler.new(
                 rate_limit: settings.sampling.rate_limit,
                 default_sample_rate: settings.sampling.default_rate
               )
             )
+          end
+        end
+
+        def ensure_priority_sampling(sampler = nil)
+          if sampler.is_a?(PrioritySampler)
+            sampler
+          else
+            PrioritySampler.new(
+              base_sampler: sampler,
+              post_sampler: Sampling::RuleSampler.new
+            )
+          end
+        end
+
+        # TODO: Writer should be a top-level component.
+        # It is currently part of the Tracer initialization
+        # process, but can take a variety of options (including
+        # a fully custom instance) that makes the Tracer
+        # initialization process complex.
+        def build_writer(settings, agent_settings)
+          if (writer = settings.tracer.writer)
+            return writer
+          end
+
+          Writer.new(agent_settings: agent_settings, **settings.tracer.writer_options)
+        end
+
+        def subscribe_to_writer_events!(writer, sampler)
+          return unless writer.respond_to?(:events) # Check if it's a custom, external writer
+
+          writer.events.after_send.subscribe(:record_environment_information, &WRITER_RECORD_ENVIRONMENT_INFORMATION_CALLBACK)
+
+          return unless sampler.is_a?(Datadog::PrioritySampler)
+
+          writer.events.after_send.subscribe(
+            :update_priority_sampler_rates,
+            &writer_update_priority_sampler_rates_callback(sampler)
           )
+        end
+
+        WRITER_RECORD_ENVIRONMENT_INFORMATION_CALLBACK = lambda do |_, responses|
+          Diagnostics::EnvironmentLogger.log!(responses)
+        end
 
-          # TODO: We reconfigure the tracer here because it has way too many
-          #       options it allows to mutate, and it's overwhelming to rewrite
-          #       tracer initialization for now. Just reconfigure using the
-          #       existing mutable #configure function. Remove when these components
-          #       are extracted.
-          tracer.configure(agent_settings: agent_settings, **build_tracer_options(settings, agent_settings))
+        # Create new lambda for writer callback,
+        # capture the current sampler in the callback closure.
+        def writer_update_priority_sampler_rates_callback(sampler)
+          lambda do |_, responses|
+            response = responses.last
 
-          tracer
```

Comment:
I really like the removal of the two-step (init + configure) initialization of the Tracer!